provider/schema/resource.go

```diff
@@ -16,47 +18,97 @@ type Resource struct {
 	// Set if this is an embedded table
 	Parent *Resource
 	// internal fields
-	table *Table
-	data  map[string]interface{}
-	id    uuid.UUID
+	table       *Table
+	data        map[string]interface{}
+	cqId        uuid.UUID
+	extraFields map[string]interface{}
+	columns     []string
 }
 
-func NewResourceData(t *Table, parent *Resource, item interface{}) *Resource {
+func NewResourceData(t *Table, parent *Resource, item interface{}, extraFields map[string]interface{}) *Resource {
 	return &Resource{
-		Item:   item,
-		Parent: parent,
-		table:  t,
-		data:   make(map[string]interface{}),
-		id:     uuid.New(),
+		Item:        item,
+		Parent:      parent,
+		table:       t,
+		data:        make(map[string]interface{}),
+		cqId:        uuid.New(),
+		columns:     getResourceColumns(t, extraFields),
+		extraFields: extraFields,
 	}
 }
 
-func (r Resource) Get(key string) interface{} {
+func (r *Resource) Get(key string) interface{} {
 	return r.data[key]
 }
 
-func (r Resource) Set(key string, value interface{}) error {
-	columnExists := funk.ContainsString(r.table.ColumnNames(), key)
+func (r *Resource) Set(key string, value interface{}) error {
+	columnExists := funk.ContainsString(r.columns, key)
 	if !columnExists {
-		return fmt.Errorf("column %s does not exits", key)
+		return fmt.Errorf("column %s does not exist", key)
 	}
 	r.data[key] = value
 	return nil
 }
 
-func (r Resource) Id() uuid.UUID {
-	return r.id
+func (r *Resource) Id() uuid.UUID {
+	return r.cqId
 }
 
-func (r Resource) Values() ([]interface{}, error) {
+func (r *Resource) Values() ([]interface{}, error) {
 	values := make([]interface{}, 0)
-	values = append(values, r.id)
+	values = append(values, r.Id())
 	for _, c := range r.table.Columns {
 		v := r.Get(c.Name)
 		if err := c.ValidateType(v); err != nil {
 			return nil, err
 		}
 		values = append(values, v)
 	}
+	for _, v := range r.extraFields {
+		values = append(values, v)
+	}
 	return values, nil
 }
+
+func (r *Resource) GenerateCQId() error {
+	if len(r.table.Options.PrimaryKeys) == 0 {
+		return nil
+	}
+	objs := make([]interface{}, len(r.table.PrimaryKeys()))
+	for i, pk := range r.table.PrimaryKeys() {
+		value := r.Get(pk)
+		if value == nil {
+			return fmt.Errorf("failed to generate cq_id for %s, pk field missing %s", r.table.Name, pk)
```

Comment:
Is it necessary to fail and exit here or do you think it may makes more sense to continue without generating the cq_id and printing out a warning message?

Reply:
I am not sure, basically this should almost never happen since a PK can't be null, if it won't fail here it will fail in the table insert